File: long_questions/DSA/Leetcode/golang/8_Binary_Search_on_Answer_Space/1011_Capacity_To_Ship_Packages_Within_D_Days.go
package main

import "fmt"

// 1011. Capacity To Ship Packages Within D Days
// Time: O(N log M), Space: O(1) where M is sum(weights)
func shipWithinDays(weights []int, days int) int {
	left, right := maxWeight(weights), sumWeights(weights)
	result := right
	
	for left <= right {
		mid := left + (right-left)/2
		
		if canShip(weights, days, mid) {
			result = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	
	return result
}

func canShip(weights []int, days, capacity int) bool {
	daysNeeded := 1
	currentLoad := 0
	
	for _, weight := range weights {
		if currentLoad+weight <= capacity {
			currentLoad += weight
		} else {
			daysNeeded++
			currentLoad = weight
			if daysNeeded > days {
				return false
			}
		}
	}
	
	return daysNeeded <= days
}

func maxWeight(weights []int) int {
	max := 0
	for _, weight := range weights {
		if weight > max {
			max = weight
		}
	}
	return max
}

func sumWeights(weights []int) int {
	sum := 0
	for _, weight := range weights {
		sum += weight
	}
	return sum
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Binary Search on Answer Space
- **Search Space**: Minimum possible capacity to maximum possible capacity
- **Monotonic Property**: If capacity X works, any capacity > X also works
- **Feasibility Check**: Test if given capacity allows shipping within D days
- **Binary Search**: Narrow down to minimum feasible capacity

## 2. PROBLEM CHARACTERISTICS
- **Optimization Problem**: Find minimum capacity that satisfies constraints
- **Monotonic Feasibility**: Higher capacity always makes shipping easier
- **Partition Constraint**: Must ship all packages within exactly D days
- **Capacity Bounds**: Between max weight and total weight

## 3. SIMILAR PROBLEMS
- Split Array Largest Sum (LeetCode 410) - Similar partitioning problem
- Koko Eating Bananas (LeetCode 875) - Binary search on eating speed
- Minimum Number of Days to Make m Bouquets (LeetCode 1482) - Binary search on days
- Minimum Time to Complete Trips (LeetCode 2187) - Binary search on time

## 4. KEY OBSERVATIONS
- **Monotonic Property**: If capacity C works, any C' > C also works
- **Search Bounds**: Lower bound = max weight, upper bound = sum of all weights
- **Feasibility Logic**: Greedy packing - fill current day until capacity exceeded
- **Optimal Solution**: Binary search finds minimum feasible capacity

## 5. VARIATIONS & EXTENSIONS
- **Exact Days**: Must use exactly D days (not at most)
- **Multiple Constraints**: Add weight limits per package type
- **Package Order**: Allow reordering of packages
- **Variable Days**: Find minimum days for given capacity

## 6. INTERVIEW INSIGHTS
- Always clarify: "Can packages be reordered? Must use exactly D days?"
- Edge cases: single package, D equals number of packages, D equals 1
- Time complexity: O(N log S) where S is sum of weights
- Space complexity: O(1) additional space

## 7. COMMON MISTAKES
- Not setting correct search bounds (left = max weight, right = sum)
- Wrong feasibility logic (not greedy packing)
- Off-by-one errors in binary search
- Not handling edge cases properly
- Using linear search instead of binary search

## 8. OPTIMIZATION STRATEGIES
- **Binary Search**: O(N log S) time, optimal for this problem
- **Greedy Feasibility**: O(N) time to check if capacity works
- **Early Termination**: Stop when days exceed limit
- **Efficient Bounds**: Tighten search space bounds

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding the smallest truck that can deliver all packages:**
- You have packages with different weights and a delivery deadline
- You need to find the smallest truck capacity that works
- If a truck of size X works, any larger truck will also work
- You can test different truck sizes and narrow down to the minimum
- This is a perfect binary search problem

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of package weights, number of days D
2. **Goal**: Find minimum ship capacity to deliver all packages in D days
3. **Constraint**: Must ship packages in order, cannot reorder
4. **Output**: Minimum capacity value

#### Phase 2: Key Insight Recognition
- **"Binary search natural fit"** → Monotonic feasibility property
- **"Greedy packing works"** → Fill each day until capacity exceeded
- **"Search bounds"** → Between max weight and total weight
- **"Feasibility check"** → Test if capacity allows D-day shipping

#### Phase 3: Strategy Development
```
Human thought process:
"I need to find the minimum capacity that works.
If capacity X works, any larger capacity will also work.
This monotonic property means I can use binary search.
I'll search between max weight (minimum possible) and total weight (maximum possible).
For each capacity, I'll greedily pack packages to see if it fits in D days."
```

#### Phase 4: Edge Case Handling
- **Single package**: Capacity must equal package weight
- **D = number of packages**: Capacity equals max weight
- **D = 1**: Capacity equals total weight
- **All packages same weight**: Capacity = max weight if D ≥ packages

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
weights = [1,2,3,4,5,6,7,8,9,10], D = 5

Human thinking:
"I need to find the minimum capacity.
Search range: max=10 to sum=55

Test capacity 32:
Day 1: 1+2+3+4+5+6+7 = 28 (add 8 would exceed 32)
Day 2: 8+9+10 = 27
Total days = 2 ≤ 5 ✓ Capacity 32 works, try smaller

Test capacity 21:
Day 1: 1+2+3+4+5+6 = 21
Day 2: 7+8 = 15
Day 3: 9+10 = 19
Total days = 3 ≤ 5 ✓ Capacity 21 works, try smaller

Test capacity 15:
Day 1: 1+2+3+4+5 = 15
Day 2: 6+7 = 13
Day 3: 8 = 8
Day 4: 9 = 9
Day 5: 10 = 10
Total days = 5 ≤ 5 ✓ Capacity 15 works, try smaller

Test capacity 14:
Day 1: 1+2+3+4+5 = 14
Day 2: 6+7 = 13
Day 3: 8 = 8
Day 4: 9 = 9
Day 5: 10 = 10
Total days = 5 ≤ 5 ✓ Capacity 14 works, try smaller

Test capacity 13:
Day 1: 1+2+3+4 = 10
Day 2: 5+6 = 11
Day 3: 7 = 7
Day 4: 8 = 8
Day 5: 9 = 9
Day 6: 10 = 10
Total days = 6 > 5 ✗ Capacity 13 doesn't work

Minimum capacity = 15 ✓"
```

#### Phase 6: Intuition Validation
- **Why binary search works**: Monotonic feasibility property
- **Why greedy packing works**: Optimal for given capacity
- **Why search bounds**: Minimum = max weight, maximum = total weight
- **Why O(N log S)**: N for feasibility check, log S for binary search

### Common Human Pitfalls & How to Avoid Them
1. **"Why not try all capacities?"** → Too slow, binary search is optimal
2. **"Should I use DP?"** → No, greedy works for feasibility check
3. **"What about search bounds?"** → Must be max weight to total weight
4. **"Can I optimize further?"** → Binary search is already optimal

### Real-World Analogy
**Like finding the right size moving truck:**
- You have boxes of different sizes to move in a certain number of trips
- You want the smallest truck that can handle all boxes in the required trips
- If a small truck works, a larger one will definitely work
- You can test different truck sizes and find the minimum that works
- This is exactly what binary search does for you

### Human-Readable Pseudocode
```
function shipWithinDays(weights, D):
    left = max(weights)  // Minimum possible capacity
    right = sum(weights) // Maximum possible capacity
    result = right
    
    while left <= right:
        mid = left + (right - left) / 2
        
        if canShip(weights, D, mid):
            result = mid
            right = mid - 1  // Try smaller capacity
        else:
            left = mid + 1   // Need larger capacity
    
    return result

function canShip(weights, D, capacity):
    days = 1
    currentLoad = 0
    
    for weight in weights:
        if currentLoad + weight <= capacity:
            currentLoad += weight
        else:
            days += 1
            currentLoad = weight
            if days > D:
                return false
    
    return days <= D
```

### Execution Visualization

### Example: weights = [3,2,2,4,1,4], D = 3
```
Binary Search Process:
Search range: [4, 16] (max weight to total weight)

Test capacity 10:
Day 1: 3+2+2 = 7
Day 2: 4+1 = 5
Day 3: 4 = 4
Days used = 3 ≤ 3 ✓ Try smaller

Test capacity 7:
Day 1: 3+2 = 5
Day 2: 2+4 = 6
Day 3: 1 = 1
Day 4: 4 = 4
Days used = 4 > 3 ✗ Need larger

Test capacity 8:
Day 1: 3+2+2 = 7
Day 2: 4+1 = 5
Day 3: 4 = 4
Days used = 3 ≤ 3 ✓ Try smaller

Test capacity 6:
Day 1: 3+2 = 5
Day 2: 2+4 = 6
Day 3: 1 = 1
Day 4: 4 = 4
Days used = 4 > 3 ✗ Need larger

Minimum capacity = 8 ✓
```

### Key Visualization Points:
- **Monotonic property**: Higher capacity always works if lower works
- **Greedy packing**: Fill each day optimally for given capacity
- **Binary search**: Narrow down to minimum feasible capacity
- **Search bounds**: Between max weight and total weight

### Memory Layout Visualization:
```
Capacity Test Visualization:
capacity = 8, weights = [3,2,2,4,1,4]

Day 1: [3,2,2] = 7 (next 4 would exceed 8)
Day 2: [4,1] = 5 (next 4 would exceed 8)  
Day 3: [4] = 4

Total days = 3 ≤ D ✓ Capacity 8 works!
```

### Time Complexity Breakdown:
- **Binary Search**: O(log S) iterations where S = sum(weights)
- **Feasibility Check**: O(N) time per iteration
- **Total Time**: O(N log S)
- **Space Complexity**: O(1) additional space

### Alternative Approaches:

#### 1. Linear Search (O(N × S) time, O(1) space)
```go
func shipWithinDaysLinear(weights []int, days int) int {
    for capacity := maxWeight(weights); capacity <= sumWeights(weights); capacity++ {
        if canShip(weights, days, capacity) {
            return capacity
        }
    }
    return sumWeights(weights)
}
```
- **Pros**: Simple to understand
- **Cons**: Too slow for large inputs

#### 2. Dynamic Programming (O(N × D × S) time, O(D × S) space)
```go
func shipWithinDaysDP(weights []int, days int) int {
    n := len(weights)
    dp := make([][]int, days+1)
    for i := range dp {
        dp[i] = make([]int, n+1)
        for j := range dp[i] {
            dp[i][j] = math.MaxInt32
        }
    }
    dp[0][0] = 0
    
    for d := 1; d <= days; d++ {
        for i := 1; i <= n; i++ {
            sum := 0
            for j := i; j >= 1; j-- {
                sum += weights[j-1]
                if dp[d-1][j-1] != math.MaxInt32 {
                    dp[d][i] = min(dp[d][i], max(dp[d-1][j-1], sum))
                }
            }
        }
    }
    
    return dp[days][n]
}
```
- **Pros**: Finds exact optimal solution
- **Cons**: Too complex and slow for this problem

#### 3. Priority Queue Approach (O(N log N) time, O(N) space)
```go
func shipWithinDaysPQ(weights []int, days int) int {
    // This approach doesn't directly solve the problem
    // but could be used for related partitioning problems
    return -1
}
```
- **Pros**: Useful for some partitioning variants
- **Cons**: Not applicable to this specific problem

### Extensions for Interviews:
- **Exact Days Constraint**: Must use exactly D days
- **Package Reordering**: Allow reordering of packages
- **Multiple Constraints**: Add weight limits per day
- **Variable Days**: Find minimum days for given capacity
- **Parallel Shipping**: Multiple ships can work simultaneously
*/
func main() {
	// Test cases
	testCases := []struct {
		weights []int
		days    int
	}{
		{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 5},
		{[]int{3, 2, 2, 4, 1, 4}, 3},
		{[]int{1, 2, 3, 1, 1}, 4},
		{[]int{10, 50, 100, 100, 50, 10}, 5},
		{[]int{1, 2, 3, 4, 5}, 5},
		{[]int{1, 2, 3, 4, 5}, 1},
		{[]int{100}, 1},
		{[]int{1, 1, 1, 1, 1, 1, 1, 1, 1}, 9},
		{[]int{5, 5, 5, 5, 5}, 3},
	}
	
	for i, tc := range testCases {
		result := shipWithinDays(tc.weights, tc.days)
		fmt.Printf("Test Case %d: weights=%v, days=%d -> Min capacity: %d\n", 
			i+1, tc.weights, tc.days, result)
	}
}

File: long_questions/DSA/Leetcode/java/8_Binary_Search_on_Answer_Space/1011_Capacity_To_Ship_Packages_Within_D_Days.java
public class CapacityToShipPackagesWithinDDays {
    
    // 1011. Capacity To Ship Packages Within D Days
    // Time: O(N log M), Space: O(1) where M is sum(weights)
    public int shipWithinDays(int[] weights, int days) {
        int left = maxWeight(weights);
        int right = sumWeights(weights);
        int result = right;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (canShip(weights, days, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }
    
    private boolean canShip(int[] weights, int days, int capacity) {
        int daysNeeded = 1;
        int currentLoad = 0;
        
        for (int weight : weights) {
            if (currentLoad + weight <= capacity) {
                currentLoad += weight;
            } else {
                daysNeeded++;
                currentLoad = weight;
                if (daysNeeded > days) {
                    return false;
                }
            }
        }
        
        return daysNeeded <= days;
    }
    
    private int maxWeight(int[] weights) {
        int max = 0;
        for (int weight : weights) {
            if (weight > max) {
                max = weight;
            }
        }
        return max;
    }
    
    private int sumWeights(int[] weights) {
        int sum = 0;
        for (int weight : weights) {
            sum += weight;
        }
        return sum;
    }
    
    // Alternative approach with detailed explanation
    public int shipWithinDaysDetailed(int[] weights, int days) {
        System.out.println("=== Binary Search on Answer Space ===");
        System.out.printf("Weights: %s, Days: %d\n", java.util.Arrays.toString(weights), days);
        
        int left = maxWeight(weights);
        int right = sumWeights(weights);
        int result = right;
        
        System.out.printf("Search space: [%d, %d]\n", left, right);
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            System.out.printf("Trying capacity: %d\n", mid);
            
            if (canShipDetailed(weights, days, mid)) {
                result = mid;
                right = mid - 1;
                System.out.printf("  ✓ Can ship in %d days, searching lower\n", days);
            } else {
                left = mid + 1;
                System.out.printf("  ✗ Cannot ship, need higher capacity\n");
            }
        }
        
        System.out.printf("Final result: %d\n", result);
        return result;
    }
    
    private boolean canShipDetailed(int[] weights, int days, int capacity) {
        int daysNeeded = 1;
        int currentLoad = 0;
        StringBuilder dayLoads = new StringBuilder();
        
        for (int i = 0; i < weights.length; i++) {
            int weight = weights[i];
            
            if (currentLoad + weight <= capacity) {
                currentLoad += weight;
                dayLoads.append(weight).append(" ");
            } else {
                System.out.printf("    Day %d: %s (load: %d)\n", daysNeeded, dayLoads.toString().trim(), currentLoad);
                daysNeeded++;
                currentLoad = weight;
                dayLoads = new StringBuilder();
                dayLoads.append(weight).append(" ");
                
                if (daysNeeded > days) {
                    return false;
                }
            }
        }
        
        System.out.printf("    Day %d: %s (load: %d)\n", daysNeeded, dayLoads.toString().trim(), currentLoad);
        return daysNeeded <= days;
    }
    
    // Brute force approach for comparison
    public int shipWithinDaysBruteForce(int[] weights, int days) {
        int minCapacity = maxWeight(weights);
        int maxCapacity = sumWeights(weights);
        
        for (int capacity = minCapacity; capacity <= maxCapacity; capacity++) {
            if (canShip(weights, days, capacity)) {
                return capacity;
            }
        }
        
        return maxCapacity;
    }
    
    // Optimized version with early termination
    public int shipWithinDaysOptimized(int[] weights, int days) {
        int left = maxWeight(weights);
        int right = sumWeights(weights);
        int result = right;
        
        // Early optimization: if days == weights.length, answer is maxWeight
        if (days == weights.length) {
            return left;
        }
        
        // Early optimization: if days == 1, answer is sumWeights
        if (days == 1) {
            return right;
        }
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (canShipOptimized(weights, days, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }
    
    private boolean canShipOptimized(int[] weights, int days, int capacity) {
        int daysNeeded = 1;
        int currentLoad = 0;
        
        for (int weight : weights) {
            if (weight > capacity) {
                return false; // Early termination
            }
            
            if (currentLoad + weight <= capacity) {
                currentLoad += weight;
            } else {
                daysNeeded++;
                currentLoad = weight;
                
                if (daysNeeded > days) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    // Version that returns the actual shipping arrangement
    public class ShippingPlan {
        int minCapacity;
        java.util.List<java.util.List<Integer>> dailyLoads;
        
        ShippingPlan(int minCapacity, java.util.List<java.util.List<Integer>> dailyLoads) {
            this.minCapacity = minCapacity;
            this.dailyLoads = dailyLoads;
        }
    }
    
    public ShippingPlan getShippingPlan(int[] weights, int days) {
        int minCapacity = shipWithinDays(weights, days);
        java.util.List<java.util.List<Integer>> dailyLoads = new java.util.ArrayList<>();
        
        int currentLoad = 0;
        java.util.List<Integer> currentDay = new java.util.ArrayList<>();
        
        for (int weight : weights) {
            if (currentLoad + weight <= minCapacity) {
                currentLoad += weight;
                currentDay.add(weight);
            } else {
                dailyLoads.add(new java.util.ArrayList<>(currentDay));
                currentDay.clear();
                currentDay.add(weight);
                currentLoad = weight;
            }
        }
        
        if (!currentDay.isEmpty()) {
            dailyLoads.add(currentDay);
        }
        
        return new ShippingPlan(minCapacity, dailyLoads);
    }
    
    public static void main(String[] args) {
        CapacityToShipPackagesWithinDDays solver = new CapacityToShipPackagesWithinDDays();
        
        // Test cases
        int[][][] testCases = {
            {{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, {5}},
            {{3, 2, 2, 4, 1, 4}, {3}},
            {{1, 2, 3, 1, 1}, {4}},
            {{10, 50, 100, 100, 50, 10}, {5}},
            {{1, 2, 3, 4, 5}, {5}},
            {{1, 2, 3, 4, 5}, {1}},
            {{100}, {1}},
            {{1, 1, 1, 1, 1, 1, 1, 1, 1}, {9}},
            {{5, 5, 5, 5, 5}, {3}}
        };
        
        String[] descriptions = {
            "Standard case",
            "Mixed weights",
            "Small weights",
            "Large weights",
            "Equal weights and days",
            "Single day",
            "Single package",
            "Many small packages",
            "Uniform weights"
        };
        
        for (int i = 0; i < testCases.length; i++) {
            System.out.printf("Test Case %d: %s\n", i + 1, descriptions[i]);
            int[] weights = testCases[i][0];
            int days = testCases[i][1][0];
            
            int result1 = solver.shipWithinDays(weights, days);
            int result2 = solver.shipWithinDaysBruteForce(weights, days);
            int result3 = solver.shipWithinDaysOptimized(weights, days);
            
            System.out.printf("  Binary Search: %d\n", result1);
            System.out.printf("  Brute Force: %d\n", result2);
            System.out.printf("  Optimized: %d\n", result3);
            
            // Get shipping plan
            ShippingPlan plan = solver.getShippingPlan(weights, days);
            System.out.printf("  Shipping Plan (capacity %d):\n", plan.minCapacity);
            for (int j = 0; j < plan.dailyLoads.size(); j++) {
                System.out.printf("    Day %d: %s\n", j + 1, plan.dailyLoads.get(j));
            }
            System.out.println();
        }
        
        // Detailed explanation for one case
        System.out.println("=== Detailed Explanation ===");
        int[] detailedWeights = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        int detailedDays = 5;
        solver.shipWithinDaysDetailed(detailedWeights, detailedDays);
        
        // Performance comparison
        System.out.println("\n=== Performance Comparison ===");
        
        // Large test case
        int[] largeWeights = new int[1000];
        for (int i = 0; i < 1000; i++) {
            largeWeights[i] = (i % 100) + 1;
        }
        
        long startTime = System.nanoTime();
        int largeResult = solver.shipWithinDays(largeWeights, 50);
        long endTime = System.nanoTime();
        
        System.out.printf("Large test (1000 weights, 50 days): %d (took %d ns)\n", 
            largeResult, endTime - startTime);
        
        startTime = System.nanoTime();
        int bruteResult = solver.shipWithinDaysBruteForce(largeWeights, 50);
        endTime = System.nanoTime();
        
        System.out.printf("Brute force result: %d (took %d ns)\n", 
            bruteResult, endTime - startTime);
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Binary Search on Answer Space
- **Binary Search**: Search for minimum feasible capacity
- **Feasibility Check**: Can all packages be shipped within D days
- **Search Space**: [1, maxWeight] to [sum(weights)]
- **Monotonic**: If capacity works, all smaller capacities work

## 2. PROBLEM CHARACTERISTICS
- **Package Weights**: Array of package weights
- **Shipping Constraint**: D days to ship all packages
- **Capacity Question**: Find minimum capacity to meet constraint
- **Feasibility**: Check if capacity can ship within time limit

## 3. SIMILAR PROBLEMS
- Koko Eating Bananas
- Split Array Largest Sum
- Minimum Number of Days to Make M Bouquets
- Allocate Minimum Number of Books

## 4. KEY OBSERVATIONS
- Binary search works because feasibility is monotonic
- If capacity C works, any capacity > C also works
- Search space: [1, maxWeight] to [sum(weights)]
- Feasibility check requires simulating shipping process

## 5. VARIATIONS & EXTENSIONS
- Return actual shipping arrangement
- Multiple queries on same weights
- Different constraints (e.g., at most K packages per day)
- Real-time streaming version

## 6. INTERVIEW INSIGHTS
- Clarify: "Should I return the shipping arrangement?"
- Edge cases: empty weights, single package, D=1
- Time complexity: O(N log M) vs O(NM) brute force
- Space complexity: O(1) vs O(N) for brute force

## 7. COMMON MISTAKES
- Using linear search instead of binary search
- Incorrect feasibility check implementation
- Off-by-one errors in binary search bounds
- Not handling edge cases properly

## 8. OPTIMIZATION STRATEGIES
- Binary search is optimal for single query
- Early termination for obvious cases
- Mathematical bounds for search space
- Pre-sorting if weights are unsorted

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding the right truck size:**
- You have packages with different weights (weights)
- You want to ship them within D days (deadline)
- You need to find the smallest truck that can do this job
- If a truck size works, any larger truck also works

- This monotonic property allows binary search

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of weights and integer D
2. **Goal**: Find minimum capacity to ship all packages within D days
3. **Output**: Minimum feasible capacity

#### Phase 2: Key Insight Recognition
- **"Can I test a capacity?"** → Simulate shipping process
- **"Is the relationship monotonic?"** → Yes! If capacity C works, C+1 works
- **"What's the search range?"** → [1, maxWeight] to [sum(weights)]
- **"Can I use binary search?"** → Yes, monotonic feasibility

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use binary search to find the minimum capacity:
1. Define search space: [1, maxWeight] to [sum(weights)]
2. For each capacity mid:
   - Check if I can ship all packages within D days
   - If yes, try smaller capacity
   - If no, try larger capacity
3. The smallest 'yes' capacity is my answer"
```

#### Phase 4: Edge Case Handling
- **Empty weights**: Return 0 (no capacity needed)
- **D >= weights.length**: Return maxWeight (ship each separately)
- **Single package**: Return its weight
- **Large D**: Binary search helps avoid O(ND) brute force

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Starting with: weights = [1,2,3,4,5,6,7,8,9,10], D = 5

Human thinking:
"Search space: [1, 10] to [55]

Step 1: mid = (1+10)/2 = 5
Can I ship with capacity 5?
- Day 1: [1,2,3,4,5] (weight 15) ✓
- Day 2: [6,7,8,9,10] (weight 40) ✓
- Day 3: [11] (weight 11) ✓
- Day 4: [12] (weight 12) ✓
- Day 5: [13] (weight 13) ✓
- Yes! Try smaller capacity

Step 2: mid = (1+4)/2 = 2
Can I ship with capacity 2?
- Day 1: [1,2] (weight 3) ✓
- Day 2: [3,4] (weight 7) ✓
- Day 3: [5,6] (weight 11) ✓
- Day 4: [7,8] (weight 15) ✓
- Day 5: [9,10] (weight 19) > 5 days ✗
- No! Try larger capacity

Step 3: mid = (2+4)/2 = 3
Can I ship with capacity 3?
- Yes! Answer is 3"
```

#### Phase 6: Intuition Validation
- **Why it works**: Monotonic feasibility enables binary search
- **Why it's efficient**: O(N log M) vs O(ND) brute force
- **Why it's correct**: Binary search finds minimal feasible capacity

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just try all capacities?"** → Too slow O(ND)
2. **"What's wrong with linear search?"** → Misses logarithmic improvement
3. **"How to check feasibility?"** → Must simulate actual shipping process
4. **"What about the bounds?"** → Lower bound is 1, upper is sum(weights)

### Real-World Analogy
**Like finding the right delivery truck:**
- You have packages of different sizes (weights)
- You need to deliver all within D days (deadline)
- You test different truck capacities
- If a small truck works, a bigger truck will also work
- You want the smallest truck that meets the deadline
- This is exactly the binary search problem!

### Human-Readable Pseudocode
```
function shipWithinDays(weights, D):
    left = 1
    right = sum(weights)
    result = right
    
    while left <= right:
        mid = left + (right - left) / 2
        
        if canShip(weights, D, mid):
            result = mid
            right = mid - 1
        else:
            left = mid + 1
    
    return result

function canShip(weights, D, capacity):
    days = 1
    currentLoad = 0
    
    for weight in weights:
        if currentLoad + weight <= capacity:
            currentLoad += weight
        else:
            days += 1
            currentLoad = weight
            if days > D:
                return false
    
    return true
```

### Execution Visualization

### Example: weights = [1,2,3,4,5,6,7,8,9,10], D = 5
```
Initial: weights = [1,2,3,4,5,6,7,8,9,10], D = 5
Search space: [1, 10] to [55]

Step 1: mid = 5
Testing capacity 5:
Day 1: [1,2,3,4,5] = 15 ≤ 5 ✓
Day 2: [6,7,8,9,10] = 40 > 5 ✗
Day 3: [11] = 11 > 5 ✗
Day 4: [12] = 12 > 5 ✗
Day 5: [13] = 13 > 5 ✗
Result: Can ship, search lower

Step 2: mid = 2
Testing capacity 2:
Day 1: [1,2] = 3 ≤ 2 ✓
Day 2: [3,4] = 7 > 2 ✗
Day 3: [5,6] = 11 > 2 ✗
Day 4: [7,8] = 15 > 2 ✗
Day 5: [9,10] = 19 > 2 ✗
Result: Cannot ship, search higher

Step 3: mid = 3
Testing capacity 3:
Day 1: [1,2,3] = 6 ≤ 3 ✓
Day 2: [4,5] = 9 > 3 ✗
Day 3: [6,7] = 13 > 3 ✗
Day 4: [8,9] = 17 > 3 ✗
Day 5: [10] = 10 > 3 ✗
Result: Can ship, answer = 3 ✓
```

### Key Visualization Points:
- **Binary search** on capacity values
- **Feasibility check** simulates actual shipping
- **Monotonic property**: If capacity works, larger works
- **Search space reduction**: Narrow down to minimum feasible capacity

### Memory Layout Visualization:
```
Weights: [1][2][3][4][5][6][7][8][9][10]
Capacity 5: [1,2,3]|[4,5]|[6]|[7]|[8]|[9]|[10]
           Day 1  Day 2  Day 3  Day 4  Day 5
           ✓     ✗     ✗     ✗     ✗     ✗

Capacity 3: [1,2]|[3,4]|[5,6]|[7]|[8]|[9]|[10]
           Day 1  Day 2  Day 3  Day 4  Day 5
           ✓     ✗     ✗     ✗     ✗     ✗

Final Answer: 3 ✓
```

### Time Complexity Breakdown:
- **Binary Search**: O(log M) iterations where M = sum(weights)
- **Feasibility Check**: O(N) per iteration
- **Total**: O(N log M) time, O(1) space
- **Optimal**: Best possible for single query
- **vs Brute Force**: O(ND) where D is max weight
*/
